ng-dev/auth/shared/node-handler.ts

```diff
@@ -0,0 +1,144 @@
+import {EventEmitter} from 'events';
+import * as Http from 'http';
+import * as Url from 'url';
+
+import {AuthorizationRequest} from '@openid/appauth';
+import {AuthorizationRequestHandler, AuthorizationRequestResponse} from '@openid/appauth';
+import {AuthorizationError, AuthorizationResponse} from '@openid/appauth';
+import {AuthorizationServiceConfiguration} from '@openid/appauth';
+import {Crypto} from '@openid/appauth';
+import {log} from '@openid/appauth';
+import {BasicQueryStringUtils, QueryStringUtils} from '@openid/appauth';
+import {NodeCrypto} from '@openid/appauth/built/node_support/crypto_utils.js';
+
+import opener from 'opener';
+
+class ServerEventsEmitter extends EventEmitter {
+  static ON_UNABLE_TO_START = 'unable_to_start';
+  static ON_AUTHORIZATION_RESPONSE = 'authorization_response';
+}
+
+export class NodeBasedHandler extends AuthorizationRequestHandler {
+  // the handle to the current authorization request
+  authorizationPromise: Promise<AuthorizationRequestResponse | null> | null = null;
+
+  /** The content for the authorization redirect response page. */
+  protected authorizationRedirectPageContent =
+    'Close your browser to continue<script>window.close()</script>';
+
+  constructor(
+    // default to port 8000
+    public httpServerPort = 8000,
+    utils: QueryStringUtils = new BasicQueryStringUtils(),
+    crypto: Crypto = new NodeCrypto(),
```

Comment:
Two questions while I see this:

* Can we have tests for all of this (even if stuff is stubbed/mocked out obviously)? is this a planned follow-up? I know such tests are pretty cumbersome to write, but it helps us feeling at least somewhat confident about common workflows. See how much we invested into release tool tests.
* Do we need these as parameters? can they just be readonly members with direct initializers?

Reply:
Sounds good to me. Should we create an empty tracking issue or something?

Reply:
Tracking issue: https://github.com/angular/dev-infra/issues/725